pkg/upgrade/upgrades/version_starvation_test.go

```diff
@@ -0,0 +1,124 @@
+	"github.com/cockroachdb/cockroach/pkg/sql/catalog/lease"
+	"github.com/cockroachdb/cockroach/pkg/testutils/testcluster"
+	"github.com/cockroachdb/cockroach/pkg/upgrade/upgradebase"
+	"github.com/cockroachdb/cockroach/pkg/upgrade/upgrades"
+	"github.com/cockroachdb/cockroach/pkg/util/leaktest"
+	"github.com/cockroachdb/cockroach/pkg/util/retry"
+	"github.com/cockroachdb/errors"
+	"github.com/stretchr/testify/require"
+)
+
+// TestLeasingClusterVersionStarvation validates that setting
+// the cluster version is done with a high priority txn and cannot
+// be pushed out. Previously, this would be normal priority and
+// get pushed by the leasing by the leasing code, leading to starvation
+// when leases were acquired with sufficiently high frequency
+// Note: This test just confirms its not normal priority by checking
+// if it can push other txns.
+func TestLeasingClusterVersionStarvation(t *testing.T) {
+	defer leaktest.AfterTest(t)()
+	ctx := context.Background()
+
+	routineChan := make(chan error)
+	waitToStartBump := make(chan struct{})
+	resumeBump := make(chan struct{})
+	clusterArgs := base.TestClusterArgs{
+		ServerArgs: base.TestServerArgs{
+			Knobs: base.TestingKnobs{
+				SQLLeaseManager: &lease.ManagerTestingKnobs{
+					LeaseStoreTestingKnobs: lease.StorageTestingKnobs{
+						ForceVersionGuardReads: true,
+					},
+				},
+				UpgradeManager: &upgradebase.TestingKnobs{
+					InterlockPausePoint:               upgradebase.AfterVersionBumpRPC,
+					InterlockReachedPausePointChannel: &waitToStartBump,
+					InterlockResumeChannel:            &resumeBump,
+				},
+				Server: &server.TestingKnobs{
+					DisableAutomaticVersionUpgrade: make(chan struct{}),
+					BinaryVersionOverride: clusterversion.ByKey(
+						clusterversion.MinSupported),
+				},
+			},
+		},
+	}
+
+	// Disable lease renewals intentionally, so that we validate
+	// no deadlock risk exists with the settings table.
+	st := clustersettings.MakeTestingClusterSettingsWithVersions(
+		clusterversion.TestingBinaryVersion,
+		clusterversion.TestingBinaryMinSupportedVersion,
+		false)
+
+	clusterArgs.ServerArgs.Settings = st
+	tc := testcluster.StartTestCluster(t, 1, clusterArgs)
+	lease.LeaseDuration.Override(ctx, &st.SV, 0)
+	lease.LeaseRenewalDuration.Override(ctx, &st.SV, 0)
+
+	defer tc.Stopper().Stop(ctx)
+	db := tc.ServerConn(0)
+	defer db.Close()
+
+	proceedWithCommit := make(chan struct{})
+	// Start a background transaction that will have an intent
+	// on the version key inside the settings table, with a
+	// normal priority (which should get pushed by the upgrade).
+	go func() {
+		<-waitToStartBump
+		tx, err := db.Begin()
+		if err != nil {
+			routineChan <- err
+			return
+		}
+		_, err = tx.Exec("SELECT name from system.settings where name='version' FOR UPDATE")
+		if err != nil {
+			routineChan <- err
+			return
+		}
+		resumeBump <- struct{}{}
+		for retry := retry.Start(retry.Options{}); retry.Next(); {
```

Comment:
I'm not sure which is the correct question to ask here: Why is this in a retry loop or why is there a `Commit` call after an infinite loop?

Are we expecting the select for update query to receive the txn retry error?